synapse/storage/registration.py

```diff
@@ -984,3 +984,256 @@ def _bg_user_threepids_grandfather_txn(txn):
         yield self._end_background_update("user_threepids_grandfather")
 
         defer.returnValue(1)
+
+    def get_threepid_validation_session(
+        self,
+        medium,
+        client_secret,
+        address=None,
+        sid=None,
+        validated=None,
+    ):
+        """Gets a session_id and last_send_attempt (if available) for a
+        client_secret/medium/(address|session_id) combo
+
+        Args:
+            medium (str): The medium of the 3PID
+            address (str): The address of the 3PID
+            sid (str): The ID of the validation session
+            client_secret (str): A unique string provided by the client to
+                help identify this validation attempt
+            validated (bool|None): Whether sessions should be filtered by
+                whether they have been validated already or not. None to
+                perform no filtering
+
+        Returns:
+            deferred {str, int}|None: A dict containing the
+                latest session_id and send_attempt count for this 3PID.
+                Otherwise None if there hasn't been a previous attempt
+        """
+        keyvalues = {
+            "medium": medium,
+            "client_secret": client_secret,
+        }
+
+        if sid:
+            keyvalues["session_id"] = sid
+        elif address:
+            keyvalues["address"] = address
+        else:
+            raise StoreError(500, "Either address or sid must be provided")
+
+        def get_threepid_validation_session_txn(txn):
+            cols_to_return = [
+                "session_id", "medium", "address",
+                "client_secret", "last_send_attempt", "validated_at",
+            ]
+
+            sql = "SELECT %s FROM threepid_validation_session" % ", ".join(cols_to_return)
+            sql += " WHERE %s" % " AND ".join("%s = ?" % k for k in iterkeys(keyvalues))
+
+            if validated is not None:
+                sql += " AND validated_at IS " + ("NOT NULL" if validated else "NULL")
+
+            sql += " LIMIT 1"
+
+            txn.execute(sql, list(keyvalues.values()))
+            row = txn.fetchone()
+
+            if not row:
+                return None
+
+            # Convert the resulting row to a dictionary
+            ret = {}
+            for i in range(len(cols_to_return)):
+                ret[cols_to_return[i]] = row[i]
```

Comment:
We have helper function that uses the returned data structure to do this for us. Instead of `txn.fetchone()`:

```python
rows = self.cursor_to_dict(txn)
if not rows:
    return None

return rows[0]
```